include/random_walks/shake_and_bake_walk.hpp

```diff
@@ -0,0 +1,169 @@
+
+struct ShakeAndBakeWalk
+{
+
+    template <typename Polytope, typename RandomNumberGenerator>
+    struct Walk
+    {
+        using Point = typename Polytope::PointType;
+        using VT = typename Polytope::VT;
+        using NT = typename Point::FT;
+        using MT = typename Polytope::MT;
+
+        struct update_parameters 
+        {
+            int   facet_prev   = -1;  
+        };
+
+        update_parameters params_;
+
+        static constexpr NT kDefaultEpsilon = NT(1e-10);
+
+        template <typename GenericPolytope>
+        Walk(GenericPolytope&          P,
+            const Point&             boundary_pt, 
+            int                      facet_idx,    
+            RandomNumberGenerator&   rng,
+            NT                       eps = kDefaultEpsilon)
+            : P_{P}, epsilon_{eps}
+        {
+            P_.normalize();
+            initialize(boundary_pt, facet_idx, rng);
+        }
+
+        void set_epsilon(NT eps) noexcept { epsilon_ = eps; }
+        NT   get_epsilon() const noexcept { return epsilon_; }
+
+         void apply(unsigned int walk_len, RandomNumberGenerator& rng)
+        {
+            const NT eps = epsilon_; 
+
+            for (unsigned step = 0; step < walk_len; ++step)
+            {
+                Point v = get_direction(rng);
+
+                int facet_new;
+                std::tie(lambda_hit_, facet_new) = P_.line_positive_intersect_skip(p_, v, Ar_, Av_, lambda_hit_, params_);
+
+                if (!std::isfinite(lambda_hit_) || lambda_hit_ <= eps  || facet_new < 0) 
+                {
+                    lambda_hit_ = NT(0);
+                    continue;
+                }
+
+                p_ += lambda_hit_ * v;
+                facet_idx_ = facet_new;
+                A_row_k_   = P_.get_facet_normal_vec(facet_idx_);
+                params_.facet_prev  = facet_idx_;
+            }
+        }
+
+
+        const Point& getCurrentPoint() const noexcept { return p_; }
+
+    private:
+
+        Point get_direction(RandomNumberGenerator& rng)
+        {
+            VT z = GetDirection<Point>::apply(dim_, rng).getCoefficients();
+            MT I_cc = - A_row_k_ * A_row_k_.transpose();
+            I_cc.diagonal() += VT::Ones(dim_);
+            NT U = rng.sample_urdist();               
+            NT r = std::pow(U, NT(1)/NT(dim_-1)); 
+            NT cz = A_row_k_.dot(z);
+            VT z_tilde  = I_cc*z;
+            z_tilde *= r;
+            z_tilde /= std::sqrt(NT(1) - cz*cz);
+            
+            VT v = z_tilde - std::sqrt(NT(1) - r*r) * A_row_k_;
+            return Point(v);
+        }
+
+        void initialize(const Point& boundary_pt,
+                        int   facet_idx,
+                        RandomNumberGenerator& rng)
+        {
+            dim_ = P_.dimension();
+            m_ = P_.num_of_hyperplanes();
+            VT b=P_.get_vec();
+
+            NT kFacetEps = NT(1e-8);
```

Comment:
Why not using the `epsilon_` here? For tolerances checks we should use the epsilon passed in the constructor.